### PR TITLE
Drain database connections on graceful shutdown

### LINKED ISSUE #21158: Database agent does not drain connections on graceful shutdown
Related issue #20126

Expected behavior:
On graceful shutdown of database agent, it would be great if the agent will wait for live/active connections to finish up to a period of time before exit.

SSH server does drain connections
https://github.com/gravitational/teleport/blob/731acf9cb731bbb540e6d2ecdd7d7297accf7d35/lib/sshutils/server.go#L349-L367

Current behavior:
Database agent exists without waiting for live/active connections.

Bug details:
- Teleport version: v12
- Recreation steps: `tsh db connect` on the client side, and `kill -QUIT <pid-of-db>` on the server side

### PR DESCRIPTION
**Pull request by gabrielcorado**
Closes #21158. Implements a similar logic as SSH server where, when on graceful shutdown, the agent waits until all database connections it is serving are closed by the clients.

This PR can be split into two parts:
1. Tracking active database connections. This is done on the `HandleConnection` function of the server.
2. Introduce a new context that is for connection-related stuff. Everything necessary for a connection to keep running (engine and audit) now relies on this context instead of the `closeContext`. In that way, we can still close resources like heartbeat without canceling the connections context.

Since the SSH server doesn't expose the shutdown polling as a configuration, I haven't done it here too. But I'm open to adding an option for users to change it.

### CODE REVIEW

**Review by greedy52 — APPROVED** [review]
LGTM. some concerns on channel check in goroutines in UT.

File: lib/srv/db/server_test.go [hunk 1]
```diff
@@ -315,3 +317,152 @@ func TestShutdown(t *testing.T) {
...
+	for i := 0; i < numActiveConnections; i++ {
+		conn, err := testCtx.postgresClient(ctx, "alice", "postgres", "postgres", "postgres")
+		require.NoError(t, err)
+		closeFuncs = append(closeFuncs, func() error { return conn.Close(ctx) })
+		// We're also adding them to the test cleanup to ensure connections are
+		// closed even when the test fails.
+		t.Cleanup(func() { conn.Close(ctx) })
+	}
+
+	require.Equal(t, int32(numActiveConnections), testCtx.server.getActiveConnections())
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/srv/db/server_test.go]: > Curious why require.Eventually is required in the count-down loop but not here

Adding to the active connections is done first and synchronously to the connection establishment, so we don't need to wait for anything. However, we could add it here as a double-checker.

> and why the count-up loop doesn't check testCtx.server.getActiveConnections() like the count-down loop.

That's a good point. I've updated the code to match both loops.

File: lib/srv/db/server.go [hunk 2]
```diff
@@ -845,11 +901,14 @@ func (s *Server) HandleConnection(conn net.Conn) {
...
 	// certificate, extract identity information from it.
-	ctx, err := s.middleware.WrapContextWithUser(s.closeContext, tlsConn)
+	ctx, err := s.middleware.WrapContextWithUser(s.connContext, tlsConn)
 	if err != nil {
 		log.WithError(err).Error("Failed to extract identity from connection.")
 		return
 	}
+	// Track active connections.
+	s.trackActiveConnections(1)
+	defer s.trackActiveConnections(-1)
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/srv/db/server.go]: You're right, it makes sense to be the first thing done on `HandleConnection`.

File: lib/srv/db/server_test.go [hunk 3]
```diff
@@ -315,3 +317,152 @@ func TestShutdown(t *testing.T) {
...
+	go func() {
+		shutdownErrCh <- server.Shutdown(shutdownCtx)
+	}()
+
+	// Shutdown doesn't return immediately. We must wait for a short period to
+	// ensure it hasn't been completed.
+	require.Never(t, func() bool {
+		<-shutdownErrCh
+		return true
+	}, time.Second, time.Second, "unexpected server shutdown")
```
[the remark this answers is not in the document]
Comment (gabrielcorado) [reply lib/srv/db/server_test.go]: I've updated the `Never` block to have a `select` where we default to `false`, triggering another tick of the check function. Also, I've updated the ticking time.

**Review by smallinsky — APPROVED** [review]

**Pull request merged**